### ISSUE TITLE
GCC package doesn't work on Linux-32

### ISSUE THREAD

**Issue by frol** [conda/conda-recipes#510]
I've tested this on CentOS 6.7 i686 and Ubuntu 12.04 i686, and GCC 4.8.5-3 (official conda repo) and GCC-5 5.3.0-0 (the one that I'm maintaining).

```
(_root)bash-4.1$ conda install gcc
Fetching package metadata: ....
Solving package specifications: ....
Package plan for installation in environment /home/jenkins/.conda/envs/_root:

The following NEW packages will be INSTALLED:

    cloog: 0.18.0-0 defaults (soft-link)
    gcc:   4.8.5-3  defaults (soft-link)
    gmp:   5.1.2-2  defaults (soft-link)
    isl:   0.12.2-0 defaults (soft-link)
    mpc:   1.0.1-0  defaults (soft-link)
    mpfr:  3.1.2-0  defaults (soft-link)

Proceed ([y]/n)?

Linking packages ...
[cloog               ]|############################################################################################                                                     
cc1: error: /root/anaconda/envs/_build/lib/gcc/i686-pc-linux-gnu/4.8.5/include: Permission denied
cc1: error: /root/anaconda/envs/_build/lib/gcc/i686-pc-linux-gnu/4.8.5/include-fixed: Permission denied
Installation failed: gcc is not able to compile a simple 'Hello, World' program.
Error: Error: post-link failed for: gcc-4.8.5-3#################################################################################################################################                               |  83%
```

I don't have `/root/anaconda/envs/_build/lib/gcc/i686-pc-linux-gnu/4.8.5/include`, it is left hard-coded into gcc binaries. This issue stems from [`detect_binary_files_with_prefix: true # [not linux32]`](https://github.com/conda/conda-recipes/blob/master/gcc-4.8/meta.yaml#L11). However, if I build the recipe with `detect_binary_files_with_prefix: true` (without `[not linux32]`), I'm getting another issue:

```
Linking packages ...
[wheel               ]|#####################################################
gcc install OS matches gcc build OS: Skipping post-link portability fixes.
#######                                                  |  73%
gcc: error: /home/jenkins/.conda/envs/_test/lib/gcc/: Is a directory
Installation failed: gcc is not able to compile a simple 'Hello, World' program.
Error: post-link failed for: gcc-4.8.5-5
##############################################################################################################################################                         |  86%
```

The interesting part here is `gcc: error: /home/jenkins/.conda/envs/_test/lib/gcc/: Is a directory`.

Here is another interesting thing:

```
$ /home/jenkins/.conda/envs/_test/bin/gcc --version
gcc: error: /home/jenkins/.conda/envs/_test/lib/gcc/: Is a directory
```

GCC cannot even print its own version; the binary is completely broken.

The package which is rebuilt without `detect_binary_files_with_prefix: true` works fine for the particular machine, but, obviously, it is not a portable package due to the hard-coded paths.

Learning how `detect_binary_files_with_prefix` -> `has_prefix` -> [`binary_replace`](https://github.com/conda/conda/blob/6c4b220326ede32f4eff92fe7010f9b1921d3fd2/conda/install.py#L250) works, and bruteforcing some ideas I came to the following result:

1. The package without `detect_binary_files_with_prefix: true` is not portable (so the official repo provides a broken package to Linux-32).
2. The package with `detect_binary_files_with_prefix: true` should work, but it doesn't. (`gcc: error: /home/jenkins/.conda/envs/_test/lib/gcc/: Is a directory`)
3. (IMPORTANT!) Merging `has_prefix` from the package **with** `detect_binary_files_with_prefix: true` (2) to the package **without** `detect_binary_files_with_prefix: true` (1), AND using a short env path (shorter than the build prefix, e.g. `/tmp/qq`) makes GCC package **work**!!! HOWEVER, not every short path works, though I couldn't find any relation, i.e. I have tried to use a path of the same length as the one that fails and it would work fine, go figure.

Here is the glitch that happens in the case (3), when the build prefix was `/home/jenkins/.conda/envs/_build/` (33 chars), and install env prefix is `/home/jenkins/.conda/envs/_root/` (32 chars):

```
Linking packages ...
gcc install OS matches gcc build OS: Skipping post-link portability fixes.
In file included from /home/jenkins/.conda/envs/_root/gcc/include/c++/bits/localefwd.h:40:0,
                 from /home/jenkins/.conda/envs/_root/gcc/include/c++/ios:41,
                 from /home/jenkins/.conda/envs/_root/gcc/include/c++/ostream:38,
                 from /home/jenkins/.conda/envs/_root/gcc/include/c++/iostream:39,
                 from hello.cpp:1:
/home/jenkins/.conda/envs/_root/gcc/include/c++/i686-pc-linux-gnu/bits/c++locale.h:52:23: error: ‘uselocale’ was not declared in this scope
   extern "C" __typeof(uselocale) __uselocale;
                       ^
/home/jenkins/.conda/envs/_root/gcc/include/c++/i686-pc-linux-gnu/bits/c++locale.h:52:45: error: invalid type in declaration before ‘;’ token
   extern "C" __typeof(uselocale) __uselocale;
                                             ^
/home/jenkins/.conda/envs/_root/gcc/include/c++/i686-pc-linux-gnu/bits/c++locale.h: In function ‘int std::__convert_from_v(__locale_struct* const&, char*, int, const char*, ...)’:
/home/jenkins/.conda/envs/_root/gcc/include/c++/i686-pc-linux-gnu/bits/c++locale.h:75:53: error: ‘__gnu_cxx::__uselocale’ cannot be used as a function
     __c_locale __old = __gnu_cxx::__uselocale(__cloc);
                                                     ^
/home/jenkins/.conda/envs/_root/gcc/include/c++/i686-pc-linux-gnu/bits/c++locale.h:100:33: error: ‘__gnu_cxx::__uselocale’ cannot be used as a function
     __gnu_cxx::__uselocale(__old);
                                 ^
Installation failed: gcc is not able to compile a simple 'Hello, World' program.
```

This thing drives me crazy!

/cc @stuarteberg @ccordoba12 @msarahan

**Comment by msarahan** [comment]
Pinging @ilanschnell - sorry, I'm not very knowledgeable here.

**Comment by stuarteberg** [comment]
FWIW, I remember testing v4.8.2 on a 32-bit Ubuntu machine, which worked. But that was building the recipe from scratch; not verifying portability.

I agree with your assessment that `detect_binary_files_with_prefix: true` should not be excluded from 32-bit builds.  I wonder why that selector was added in the first place...

Side question: are you able to get a build the package at all using `conda build --no-test gcc-4.8`?

>using a short env path (shorter than the build prefix, e.g. /tmp/qq) makes GCC package work

OK, I think you probably already know this, but just to be clear: The need for a short prefix is probably only necessary because you had to hack together a package from a build that didn't use `detect_binary_files_with_prefix: true` in the first place.  When `detect_binary_files_with_prefix: true` is used, I think `conda` intentionally uses a really really long build prefix to avoid exactly this problem.

If you grep for `/home/jenkins/.conda/envs/_test/lib/gcc/` in the binary files of your package, do you find any files that are missing from `has_prefix`?

**Comment by frol** [comment]
@stuarteberg 

> The need for a short prefix is probably only necessary because you had to hack together a package from a build that didn't use `detect_binary_files_with_prefix: true` in the first place. When `detect_binary_files_with_prefix: true` is used, I think conda intentionally uses a really really long build prefix to avoid exactly this problem.

Yes, I do understand that. There are a number of weirdness with this package: the package with `detect_binary_files_with_prefix: true` should work in the first place, but patching the package that has a long build path doesn't work for me at all, while patching the binaries that were built with a shorter build path works *(in some cases)*.

> If you grep for `/home/jenkins/.conda/envs/_test/lib/gcc` in the binary files of your package, do you find any files that are missing from `has_prefix`?

I don't really get the idea here. Do you suggest to grep something like this:

```
$ grep -R '/home/jenkins/.conda/envs/_test/lib/gcc' /home/jenkins/.conda/envs/.pkgs/gcc-4.8.5-5/
```

? I don't see any point in this, but I did the grep and nothing was found. I have also grepped for the build prefix in the package binaries and all the binaries are included in `has_prefix`.

**Comment by stuarteberg** [comment]
>>If you grep for `/home/jenkins/.conda/envs/_test/lib/gcc` in the binary files of your package, do you find any files that are missing from `has_prefix`?

>I don't really get the idea here. Do you suggest to grep something like this:

>     $ grep -R '/home/jenkins/.conda/envs/_test/lib/gcc' /home/jenkins/.conda/envs/.pkgs/gcc-4.8.5-5/

Sorry, I wasn't clear, and furthermore, I think I was suggesting the wrong string to grep for.  I was wondering about the *installed* copy of the package files, not the copy in `pkgs`.  So, using your procedure for hacking together a package that works, and supposing it is installed to your `_test` environment, I guess this is what I was thinking:

    $ cd `/home/jenkins/.conda/envs/_test`
    $ grep -l -R $(pwd) . | cut -c 3- | sort > actual-references-to-lib-gcc
    $ cut -f3 -d ' ' /home/jenkins/.conda/pkgs/gcc-4.8.5-5/info/has_prefix | sort > conda-references-to-lib-gcc
    $ diff actual-references-to-lib-gcc conda-references-to-lib-gcc

If that diff result is non-empty (except for non-gcc files like `libcloog-isl.la`, `libgmp.la`, etc.), then something went wrong in the `detect_binary_files_with_prefix` process.  Honestly, I don't know if that will be a waste of time to try... I'm just as puzzled as you are about what could be going wrong here...

**Comment by connie** [comment]
I'm now getting the same error when trying to install gcc for linux-32.  Found that the 4.8.2 version on asmeurer's channel on anaconda works.  However, uploading it to my own channel and installing using 

conda install -c rmg gcc I always end up being prompted to install 4.8.5.  Was there some change to anaconda that overrides the channel provided package with the latest available version from defaults?

**Comment by stuarteberg** [comment]
>Was there some change to anaconda that overrides the channel provided package with the latest available version from defaults?

I think this has always been the case (so, no change from previous behavior).  If I understand correctly, adding the `-c` flag merely adds a channel to the set of available channels (including `defaults`), and then the "best" package from any all available channels is downloaded.  Since the `defaults` channel already has gcc `4.8.5`, then it is always preferred over the `4.8.2` version in your channel.  I suppose you could try removing `defaults` from your `.condarc` file as a temporary workaround.

**Comment by fonnesbeck** [comment]
Any resolution on this? I cannot get gcc to install on linux-32 from conda either.

**Comment by msarahan** [comment]
Not on this in particular, but we are working on a docker build image here: https://github.com/ContinuumIO/docker-images/pull/20

To use it, install docker (toolbox for mac/windows), put docker_wrapper.sh and either or both of the start_* scripts in a folder, and run one of the start_* scripts (you probably want the C++98 one for now).

That has multilib capabilities.  To compile 32-bit binaries with it, you can either set CFLAGS="${CFLAGS} -m32", or set CONDA_FORCE_32BIT=1, which does the same thing.  The former might be better for now, I'm not sure the latter is in the current conda-build release.

If you compile C++ code, you'll need to ship the libstdc++ that matches.  Right now, I only have the C++98 files: https://anaconda.org/msarahan/libgcc/files

**Comment by fonnesbeck** [comment]
Will try, thanks. I assume I also have to change the following?

    cp `conda info --root`/conda-bld/linux-64/* /output

**Comment by msarahan** [comment]
where are you seeing that?  If that's part of a recipe, it should definitely be changed to be less strictly hardcoded to platform.

**Comment by fonnesbeck** [comment]
Oh, I see. I was looking in master, and not on your branch. Thinking about it some more, this is for continuous integration of the master branch of a package, so I should probably hold off on using a WIP at this stage.

**Comment by mingwandroid** [comment]
I'd welcome testing of https://anaconda.org/rdonnellyr/gcc . I fixed the core bug here, but disagree with using a separate GCC_PREFIX which caused it in the first place.

**Comment by stuarteberg** [comment]
@mingwandroid:

>I'd welcome testing of https://anaconda.org/rdonnellyr/gcc

Are the corresponding recipe files on github somewhere?  I'm curious to see the changes you made.

>disagree with using a separate `GCC_PREFIX`

Just in case you haven't seen it yet, `GCC_PREFIX` was introduced via #401.  That PR fixed a critical problem, so we can't get rid of it without implementing an alternative solution.

I can't think of an alternative fix at the moment, but that doesn't mean an alternative fix is impossible. :-)  For example, maybe some sort of magic could be added to the gcc `specs` file to fix this issue?  Not sure.

**Comment by mingwandroid** [comment]
I didn't see it before now, thanks @stuarteberg.

My recipe and patch are here: https://github.com/mingwandroid/conda-recipes/tree/master/gcc-4.8

The thing is, for the previous plan to work we are actually relying on GCC failing to relocate `$PREFIX/gcc/include` and that's why you see:
```
    cc1: error: /root/anaconda/envs/_build/lib/gcc/i686-pc-linux-gnu/4.8.5/include: Permission denied
```
.. whoever build the last i686 Linux GCC built it as the `root` user in their home directory. The x86_64 one on the other hand fails to find the folder without this permissions problem and just doesn't relocate the sysroot include path, dropping it but reporting it as a non-existent path (unless it happens to exist).

If your goal, as I think it is, is to remove the sysroot include path altogether, then you need to modify `gcc/incpath.c`.

I still disagree with this goal. Conda's `PREFIX` at build-time is exactly a sysroot, and sysroot include folders are meant to appear after `/usr/local/include` and before `/usr/include` but most importantly, *after* GCC and libstdc++'s internal headers in the search order. GCC depends on this for `include_next` to work for example.

I think if people have headers in `/usr/local/include` that are breaking things, then they need to not do that. I understand that `homebrew` on OS X's unfortunate decision to use this folder might cause a lot of breakage on that platform.

I wouldn't be opposed however, to removing `/usr/local/include` from GCCs include search paths on OS X due to the popularity of `homebrew`, but I don't think it should be done on other platforms since some people might prefer to iterate on their own libraries in that folder and will expect conda's GCC to pick those headers and libraries in preference to those in the conda prefix. That is, after all what the GCC developers had in mind (as best I can fathom it).

**Comment by frol** [comment]
@mingwandroid Your packages work like a charm both on CentOS 6 x86 and x86_64! I'm sorry it took so long to test them out.

However, `lsb_release` command is not packaged by default in CentOS and requires to install `redhat-lsb-core` package, which depends on `perl` and other quite heavy dependencies which in total requires to download ~55MB of packages which take ~128MB on disk. Thus, It would suggest to revert the change introducing `lsb_release -a` instead of a simple `cat /etc/*-release`.

**Comment by mingwandroid** [comment]
Ok I'll implement your suggestion.

I'm still dubious about the need for/correctness of the `GCC_ PREFIX` stuff that I made work and would welcome your opinion on that.

**Comment by frol** [comment]
Yeah, I remember the time when `GCC_PREFIX` was introduced and as @stuarteberg pointed out in #401:

> Now the user can't control the include order properly. In my case, an old version of a library (say, hdf5) happened to be installed in `/usr/local/include`. Ordinarily, it shouldn't cause any problems because I passed `-I $PREFIX/include` to the compiler, so my hdf5 headers in `$PREFIX/include` should be found first and used in my program. But since gcc thinks `$PREFIX/include` is a system header, my `-I` argument is ignored, and that directory is searched last. The outdated hdf5 headers in `/usr/local/include` were found first, and my recipe didn't compile properly.

However, reading the description of the problem, I would think of a solution where `-I` will be "fixed" by not breaking the paths order in the first place in GCC instead of dancing around and placing GCC prefix in a custom folder.

**Comment by mingwandroid** [comment]
Having our of date stuff in `/usr/local/include` is user error IMHO and there could be users who *want* that folder searched in. That was the point if it after all.

I will concede that homebrew on OS X using that folder breaks the normal usage scenario, so I would probably modify OS X GCC to not search there. If people feel strongly enough that we shouldn't look in there on Linux too then fair enough, I'd remove it there too as it's not that big a deal.

**Comment by frol** [comment]
@mingwandroid The problem was that if you specify the following order:

```
$ gcc -I$PREFIX/include -I/usr/local/include -I/opt/include -I/anything/else/include ...
```

The conda packaged GCC without `$GCC_PREFIX` workaround will ignore `-I$PREFIX/include` since it is "a system include path", and will search it only after other `-I` paths. @stuarteberg placed GCC into its own prefix to force GCC respecting `-I$PREFIX/include` as a valid "user-customized" include path.

**Comment by frol** [comment]
FYI, I'm maintaining `gcc-5` (https://github.com/salford-systems/conda-recipes/tree/master/gcc-5) and `gcc-6` (https://github.com/salford-systems/conda-recipes/tree/master/gcc-6) packages for the projects we build at Salford Systems and I has successfully applied your patches and `gcc-6` package has just finished building (`linux-32`, `linux-64`) and everyone can try it:

```
$ conda install -c conda-forge -c salford_systems gcc-6
```

Later today new `gcc-5` should be also available in `salford_systems` repo.

OS X packaging is broken, though. To be fair, there was just one single attempt when it succeeded on my build server as is in this repo, afterwards I had to apply some workarounds (adding `,-rpath,${PREFIX}/lib` to `LDFLAGS` and splitting make commands manually into 6 stages). However, your changes introduced the following error in conda-build after GCC is finaly built on OS X:

```
Fixing permissions
updating shebang: bin/aot-compile
Fixing linking of /usr/lib/libiconv.2.dylib in /Users/jenkins/miniconda3/envs/_build/bin/c++
Fixing linking of /Users/jenkins/miniconda3/envs/_build/gcc/../lib/libstdc++.6.dylib in /Users/jenkins/miniconda3/envs/_build/bin/c++
Error: Could not find gcc/../lib/libstdc++.6.dylib
```

Do you have an idea of how would I fix this?

**Comment by frol** [comment]
Oh, the issue with the library paths seems to be fixed in conda-build 1.21.x, but I cannot upgrade it since it requires conda >=4.1, which is buggy, and conda 4.2.0 is also broken for conda-build imports... I hope 4.2.1 will be released soon and it won't be a piece of crap (am I naive?).

**Comment by mingwandroid** [comment]
I know you are maintaining `gcc-5` packages but it would be really good if you could build them on `CentOS 5.11` instead of `CentOS 6`. That way Continuum staff could help to test them. FWIW, I build GCC 6.1 on CentOS 5.11 without any problems.

I'm sorry to hear `conda` hasn't been performing to your satisfaction recently. I would hope 4.2.1 is meeting your expectations.

Have you got some references for the bugs that are preventing you from using `conda 4.1.x`? If those bugs have been fixed in `4.2.x` then backporting the fixes could be considered I guess.

**Comment by frol** [comment]
@mingwandroid I believe the bugs are fixed in 4.2.x, so I don't see a point in jumping into backporting now. I have manually patched my conda_build package on the build server so GCC just can be built now.

I'm using CentOS 6 because some of the proprietary tools we use don't work with CentOS 5.11, and maintaining a separate environment for GCC is not a priority for us at this point of time, though I also used to compile early GCC 5.x on CentOS 5.11 with no issues.

**Comment by mingwandroid** [comment]
> @mingwandroid The problem was that if you specify the following order:

> `$ gcc -I$PREFIX/include -I/usr/local/include -I/opt/include -I/anything/else/include ...`

>The conda packaged GCC without $GCC_PREFIX workaround will ignore -I$PREFIX/include since it is "a system include path", and will search it only after other -I paths. @stuarteberg placed GCC into its own prefix to force GCC respecting -I$PREFIX/include as a valid "user-customized" include path.

Thanks, I'm pretty aware of what was changed and the logic behind it (see my detailed in-line comments in https://github.com/conda/conda-recipes/commit/853f506ca2a3fe8f763428dd645c7a70da380aa8).

Fudging GCC into a different prefix so that the conda-sysroot is no longer considered a sysroot is not the correct approach. From the discussions we've had and the commits I've seen, essentially, this has been done so that they are found before `/usr/local/include` and for no other reason. If we really don't want that folder to be searched in before the conda-sysroot, then we probably don't want it searched in at all and in that case we should remove it from the search-paths list and be happy.

Regarding CentOS 5.11 and GCC, it'd be good if you could reconsider. IMHO you are limiting your audience unnecessarily but I understand if you just haven't got the bandwidth.

**Comment by stuarteberg** [comment]
>Fudging GCC into a different prefix so that the conda-sysroot is no longer considered a sysroot is not the correct approach.

Ah, I think this may be the biggest discussion point.  Conceptually, is conda's `$PREFIX/include` a "system" include directory or not?  I thought the answer was "no".

>essentially, this has been done so that they are found before `/usr/local/include` and for no other reason.

Is that the only directory that could matter here? (Really asking.)  In general, the user expects to be able to control the include path order via `-I` flags.  Disregarding that order leads to confusion and can be very tricky to debug.  I'm worried about disobeying the users expectations, even if we'd be "helping him out" by doing "what he *really* wants".

>If we really don't want that folder to be searched in before the conda-sysroot, then we probably don't want it searched in at all

That might be a viable solution.  I'm a little nervous about it, though. 😟

**Comment by mingwandroid** [comment]
> Ah, I think this may be the biggest discussion point. Conceptually, is conda's $PREFIX/include a "system" include directory or not? I thought the answer was "no".

I can see both viewpoints, but I veer towards it being closer to a normal sysroot than not. This matters a lot for things like GCCs `include_next` directive and some of the more esoteric platforms (mingw-w64) are likely to not work at all otherwise (ok, we get our mingw-w64 GCC from MSYS2, so this isn't an immediate concern). Another suspicion I have it that I don't *think* we could build and use our own standard C++ library (llvm's libc++ for example) with the current ordering.

> That might be a viable solution. I'm a little nervous about it, though. :worried:

Understood. I didn't have complete courage in my convictions as I didn't change it - or more accurately, I only untangled the situation in my mind once I'd written the fix, and then had to move on to other things.

I want to switch our compiler builds to being built with `crosstool-ng` ideally.

**Comment by frol** [comment]
@mingwandroid I have just hit into an issue when using Intel C++ Compiler with the patched GCC with `detect_binary_files_with_prefix: false` (it is 5.4.0-2 on salford_systems channel) results in the following error message:

```
$ icpc qq.cpp
icpc: error #10001: could not find directory in which the set of libstdc++ include files resides
```

Using `strace` I have received the following output near the error:

```
stat("=/usr/local/include", 0x7ffe8b4615a0) = -1 ENOENT (No such file or directory)
stat("/usr/local/include", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
stat("=/usr/local/miniconda/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.4.0/include", 0x7ffe8b4615a0) = -1 ENOENT (No such file or directory)
stat("/usr/local/miniconda/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.4.0/include", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
stat("=/usr/local/miniconda/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.4.0/include-fixed", 0x7ffe8b4615a0) = -1 ENOENT (No such file or directory)
stat("/usr/local/miniconda/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.4.0/include-fixed", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
stat("=/usr/local/miniconda/bin/../include", 0x7ffe8b4615a0) = -1 ENOENT (No such file or directory)
stat("/usr/local/miniconda/bin/../include", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
stat("=/usr/include", 0x7ffe8b4615a0)   = -1 ENOENT (No such file or directory)
stat("/usr/include", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
stat("/usr/local/miniconda/bin/../include/c++/5.4.0", 0x7ffe8b4615a0) = -1 ENOENT (No such file or directory)
stat("/usr/local/miniconda/bin/../x86_64-unknown-linux-gnu/include/c++/5.4.0", 0x7ffe8b4615a0) = -1 ENOENT (No such file or directory)
open("/usr/local/bin/g++", O_RDONLY)    = 3
close(3)                                = 0
open("/tmp/icpcdashv4YPFMW", O_RDONLY)  = 3
fstat(3, {st_mode=S_IFREG|0600, st_size=3255, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fd5d19df000
read(3, "Using built-in specs.\nCOLLECT_GC"..., 4096) = 3255
read(3, "", 4096)                       = 0
stat("/home/jenkins/.conda/envs/_build/gcc/home/jenkins/.conda/envs/_build/gcc/include/c++", 0x7ffe8b441560) = -1 ENOENT (No such file or directory)
stat("/home/jenkins/.conda/envs/_build/gcc/include/c++", 0x7ffe8b441560) = -1 ENOENT (No such file or directory)
stat("/usr/local/miniconda/bin/../home/jenkins/.conda/envs/_build/gcc/include/c++", 0x7ffe8b441560) = -1 ENOENT (No such file or directory)
close(3)                                = 0
munmap(0x7fd5d19df000, 4096)            = 0
stat("/usr/include/g++", 0x7ffe8b4615a0) = -1 ENOENT (No such file or directory)
fstat(2, {st_mode=S_IFCHR|0600, st_rdev=makedev(136, 4), ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fd5d19df000
write(2, "icpc: error #10001: could not fi"..., 97icpc: error #10001: could not find directory in which the set of libstdc++ include files resides
) = 97
```

As you may see there are weirdly crafted full paths with the location where GCC was "installed" during `conda build`:

```
stat("/home/jenkins/.conda/envs/_build/gcc/home/jenkins/.conda/envs/_build/gcc/include/c++", 0x7ffe8b441560) = -1 ENOENT (No such file or directory)
stat("/home/jenkins/.conda/envs/_build/gcc/include/c++", 0x7ffe8b441560) = -1 ENOENT (No such file or directory)
stat("/usr/local/miniconda/bin/../home/jenkins/.conda/envs/_build/gcc/include/c++", 0x7ffe8b441560) = -1 ENOENT (No such file or directory)
```

Here is what I get grepping:

```
$ grep -r '/home/jenkins/.conda/envs/_build/gcc/include/c++' /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/

Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/gcc/libexec/gcc/x86_64-unknown-linux-gnu/5.4.0/jc1 matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/gcc/libexec/gcc/x86_64-unknown-linux-gnu/5.4.0/f951 matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/gcc/libexec/gcc/x86_64-unknown-linux-gnu/5.4.0/cc1obj matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/gcc/libexec/gcc/x86_64-unknown-linux-gnu/5.4.0/cc1plus matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/gcc/libexec/gcc/x86_64-unknown-linux-gnu/5.4.0/cc1 matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/gcc/libexec/gcc/x86_64-unknown-linux-gnu/5.4.0/lto1 matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/bin/x86_64-unknown-linux-gnu-gcc-5.4.0 matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/bin/x86_64-unknown-linux-gnu-c++ matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/bin/gfortran matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/bin/x86_64-unknown-linux-gnu-g++ matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/bin/c++ matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/bin/gcj matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/bin/x86_64-unknown-linux-gnu-gcc matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/bin/gcc matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/bin/g++ matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/bin/x86_64-unknown-linux-gnu-gcj matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/bin/cpp matches
Binary file /usr/local/miniconda/pkgs/gcc-5-5.4.0-2/bin/x86_64-unknown-linux-gnu-gfortran matches
```

Downgrading to the previous version without the patch and with `detect_binary_files_with_prefix: true` (GCC 5.4.0-1) works just fine (though, obviously, the mentioned problems with portability might have arisen, but since I run GCC on the same infrastructure as I had built GCC on, it works fine).

Do you have any idea on how I can solve it? Are you sure that `detect_binary_files_with_prefix: true` is a bad thing?
